Sync slider with active keyframe

Fifourche opened the pull request:
A PR to sync the slider with the active keyframe, so that if it is changed programmatically or via widget selection, the slider's value is changed accordingly !

alisterburt reviewed (changes requested):
this is great @Fifourche! love it!

There's a small bug, when moving the slider slowly backwards you need to make sure to change the selection without modifying the slider position- see video for what I mean

https://user-images.githubusercontent.com/7307488/119902307-f7ba6e80-bf3e-11eb-90b9-7d49c2d07f47.mp4

We should probably add a test for that - once this is fixed let's get this merged 😄

alisterburt commented:
awesome! could you look into the test which is failing before we merge? https://github.com/napari/napari-animation/pull/93/checks?check_run_id=2692537547#step:9:233

Fifourche commented:
Yes I'll have time to fix it soon ! :)

Fifourche commented:
I commited some changes - too many at once probably ! I'll avoid that in the future, for the git history to be more readable.

- In here I removed the `current_key_frame attribute`, and the references to `key_frames.selection._current` to rely only on `key_frames.selection.active` ; as we discussed it elsewhere, only one of them would be needed and `keyframe.selection.active` corresponds to the selected keyframe in GUI !

- Fixed the `slider_moved` callback issue. 

- If no frame is selected (active), errors are raised when attempting deletion or replacement, and delete button is disabled.

alisterburt commented on napari_animation/_qt/animation_widget.py:
```diff
@@ -122,6 +126,30 @@ def _on_keyframes_changed(self, event=None):
+            bool(active_keyframe)
+        )
+
+    def _slider_moved(self, event=None):
```
could you rename this `_on_slider_moved` to be consistent with the other methods?

alisterburt commented on napari_animation/animation.py:
```diff
@@ -90,23 +99,18 @@ def _validate_animation(self):
-    @property
-    def current_key_frame(self):
-        return self.key_frames.selection._current
+        key_frame = self._frames._frame_index[index][0]
+        self.key_frames.selection.active = key_frame
```
really nice!

alisterburt reviewed (commented):
Great stuff @Fifourche - I suggested one small change (and put a PR for it at fifourche/napari-animation#4), merge that if you're happy then we're good to go!

Fifourche commented:
Done ! Thanks again for the help and the good vibes :)

alisterburt commented:
great working with you as always!
